Add some troubleshooting instructions.

brendandburns opened the pull request:
@kprobst @juliaferraioli @dchen1107 @MikeJeffrey @RichieEscarez

brendandburns committed:
Add some trouble shooting instructions.

derekwaynecarr commented:
nice write-up.

derekwaynecarr commented:
One thing I wondered if we need a section on is debugging why your image cannot be pulled.

dchen1107 commented:
Nice! 

Several things we can add (not necessary in this PR):

1) Why pod is pending? 
Scheduling? resource capacity? image issue? 

2) How to debugging network / dns related issue?  iptables-save, etcdctl ls those srv records

3) How to monitor resource stats related to a node, pod or container?

juliaferraioli commented:
Looking good! I'd also link to the documentation for various concepts the first time you mention them, like for pods, replication controller, service, etc...

kprobst commented on docs/application-troubleshooting.md:
```diff
@@ -0,0 +1,104 @@
+# Application Troubleshooting.
+
+This guide is to help users debug applications that are deployed into Kubernetes and not behaving correctly.
+This is *not* the guide for people who want to debug their cluster.  For that you should check out
```
suggest "the" -> "a"

brendandburns replied:
done.

kprobst commented on docs/application-troubleshooting.md:
```diff
@@ -0,0 +1,104 @@
+[this guide](cluster-troubleshooting.md)
+
+## FAQ
+Users are highly encouraged to checkout our [FAQ](https://github.com/GoogleCloudPlatform/kubernetes/wiki/User-FAQ)
```
nit: I'm pretty sure this should be "check out" instead of "checkout".

brendandburns replied:
done.

kprobst commented on docs/application-troubleshooting.md:
````diff
@@ -0,0 +1,104 @@
+```
+
+If your container has previously crashed, you can access the previous containers crash log with:
+```sh
````
previous container's

brendandburns replied:
done.

kprobst commented on docs/application-troubleshooting.md:
````diff
@@ -0,0 +1,104 @@
+
+```sh
+kubectl exec ${POD_NAME} -c ${CONTAINER_NAME} -- ${CMD} ${ARG1} ${ARG2} ... ${ARGN}
+```
````
It would be good to give a specific example here, or point people to documentation for what commands are availabe here.

brendandburns replied:
done.

kprobst commented on docs/application-troubleshooting.md:
```diff
@@ -0,0 +1,104 @@
+
+If none of these approaches work, you can find the host machine that the pod is running on and SSH into that host,
+but this is generally not recommended, and indeed if you find yourself needing to ssh into a machine, please file a
+feature request on GitHub describing your use case and why these tools are insufficient.
```
Nit: Suggest splitting this into two sentences.

Also, why is this not recommended? How about instead we say something like "this shouldn't be needed because of all the other awesome debugging tools we give you"?

brendandburns replied:
done.

kprobst commented on docs/application-troubleshooting.md:
````diff
@@ -0,0 +1,104 @@
+```
+
+Make sure that the endpoints match up with your expectations.
+
````
Hmm..if I'm reading a troubleshooting doc on how to debug services, my guess is I don't know what expectations I should have of the endpoints. Can you be a little more specific here?

brendandburns replied:
done.

kprobst commented on docs/application-troubleshooting.md:
````diff
@@ -0,0 +1,104 @@
+kubectl get pods --selector=name=nginx,type=frontend
+```
+
+To list pods that match this selector.  Verify that the list matches your expectations.
````
Nit: To -> to

Same comment about expectations.

brendandburns replied:
done.

kprobst commented on docs/application-troubleshooting.md:
```diff
@@ -0,0 +1,104 @@
+
+#### Network traffic isn't forwarded
+If you can connect to the service, but the connection is immediately dropped, but there are endpoints
+in the endpoints list, it's likely that the proxy can't contact your pods.  There are three things to
```
Nit: instead of the second "but" consider "and"

brendandburns replied:
done.

kprobst commented on docs/application-troubleshooting.md:
````diff
@@ -0,0 +1,104 @@
+check:
+   * Are your pods working correctly?  Look for restart count, and [debug pods](#debugging-pods)
+   * Can you connect to your pods directly?  Get the IP address for the Pod, and try to connect directly to that IP
+   * Is your application serving on the port that you configured?  Remember, Kubernetes doesn't do port remapping, so if your application serves on 8080, the ```containerPort``` field needs to be 8080.
````
Consider dropping the "remember" (this will be newbie users mostly)

brendandburns replied:
done.

kprobst commented on docs/cluster-troubleshooting.md:
```diff
@@ -0,0 +1,27 @@
+# Cluster Trouble Shooting
```
Trouble Shooting -> Troubleshooting

brendandburns replied:
done.

kprobst commented on docs/troubleshooting.md:
```diff
@@ -0,0 +1,4 @@
+# Troubleshooting
+Things go wrong.  This guide is aimed at making them right.  It has two sections:
```
To make this a little more positive, how about "Sometimes things go wrong"?

brendandburns replied:
done.

kprobst commented:
Overall, this looks great, I've left a bunch of some small comments.

Consider adding the sections others suggest or file a bug for those so we don't lose track of them.

brendandburns commented:
I addressed @kprobst comments, as well as @derekwaynecarr 's comment.

I addressed the first of @dchen1107 's comments, and added issues (#9552, #9553) to track the other two.

I'm going to punt on @juliaferraioli , since that is a broader TODO for the docs in general, and I plan on writing some code to do that automagically.

kprobst commented:
Looks good for my comments.
